### PR TITLE
[16.0] partner_delivery_windows: Add helper function to find next delivery w…

### PR DESCRIPTION
**Pull request by grindtildeath**
…indows

### CODE REVIEW

File: stock_partner_delivery_window/models/res_partner.py [hunk 1]
```diff
@@ -180,3 +183,76 @@ def copy_data(self, default=None):
...
+        If from_date is already in a delivery window, from_date is "the next"
+
+        :param from_date: Datetime object (Leave empty to use now())
+        :param timedelta_days: Number of days to use in the computation
+                               (Leave empty to use 7 days or 1 week)
+        :return: Datetime object
+        """
+        self.ensure_one()
+        if not from_date:
+            from_date = datetime.now()
```
Comment (simahawk) [inline stock_partner_delivery_window/models/res_partner.py]:
```suggestion
        from_date = from_date or datetime.now()
```

File: stock_partner_delivery_window/models/res_partner.py [hunk 2]
```diff
@@ -180,3 +183,76 @@ def copy_data(self, default=None):
...
+        if self.is_in_delivery_window(from_date):
+            return fields.Datetime.to_datetime(from_date)
+        if timedelta_days is None:
+            timedelta_days = 7
```
Comment (simahawk) [inline stock_partner_delivery_window/models/res_partner.py:204]:
```suggestion
        timedelta_days = timedelta_days or 7
```
unless you support `0`.

File: stock_partner_delivery_window/models/res_partner.py [hunk 3]
```diff
@@ -180,3 +183,76 @@ def copy_data(self, default=None):
...
+        :param from_datetime: Datetime object
+        :param to_datetime: Datetime object
+        :return: List of Datetime objects
+        """
+        res = list()
+        for this_datetime in date_range(from_datetime, to_datetime, timedelta(days=1)):
+            this_weekday_number = this_datetime.weekday()
+            this_weekday = self.env["time.weekday"].search(
+                [("name", "=", this_weekday_number)], limit=1
+            )
```
Comment (simahawk) [inline stock_partner_delivery_window/models/res_partner.py]: When the date_range is big you'll have many many searches.
I'd recommend doing the following:

```suggestion
        dts = date_range(from_datetime, to_datetime, timedelta(days=1))
        weekdays = {x.name: x for x in self.env["time.weekday"].search([("name", "in", dts)])}
        for this_datetime, this_weekday in weekdays.items():
            this_weekday_number = this_datetime.weekday()
```

**Review by simahawk — COMMENTED** [review]
any test?

File: stock_partner_delivery_window/models/res_partner.py (hunk 2 above)
Comment (grindtildeath) [reply stock_partner_delivery_window/models/res_partner.py:204]: IMO it's usually safer to consider 0 when working on dates/times and deltas

**Comment by grindtildeath** [comment]
@mmequignon since you're the author of the main commit here, did you write any tests?

**Comment by mmequignon** [comment]
> @mmequignon since you're the author of the main commit here, did you write any tests?

be my guest https://github.com/OCA/sale-workflow/blob/4fd856f263a696f16814834e2ffd0e89bc230308/sale_delivery_date/tests/test_sale_partner_delivery_window.py

**Review by jbaudoux — APPROVED** [review]
LG

**Comment by jbaudoux** [comment]
@grindtildeath pre-commit to fix

File: stock_partner_delivery_window/models/res_partner.py [hunk 4]
```diff
@@ -180,3 +183,114 @@ def copy_data(self, default=None):
...
+
+        :param from_datetime: Datetime object
+        :param to_datetime: Datetime object
+        """
+        self.ensure_one()
+        if to_datetime is None:
+            to_datetime = from_datetime + timedelta(days=365)
+        for this_datetime in date_range(from_datetime, to_datetime, timedelta(days=1)):
+            if self.delivery_time_preference == "anytime":
+                yield this_datetime
```
Comment (jbaudoux) [inline stock_partner_delivery_window/models/res_partner.py]: I haven't look your usage yet but I think your generator should allow a new from_datetime on each iteration.
Then you can ask multiple modules the next availability and send them the new start date to validate 
Something like:
```
this_datetime = from_datetime
while True:
  if self.delivery_time_preference == "anytime":
    from_datetime = (yield this_datetime)
  elif self.delivery_time_preference == "workdays" and this_weekday_number < 5:
    from_datetime = (yield this_datetime)
  else:
    ...

  if from_datetime is not None:
    this_datetime = from_datetime
  else:
    this_datetime += timedelta(days=1)
```

File: stock_partner_delivery_window/models/res_partner.py [hunk 5]
```diff
@@ -180,3 +183,114 @@ def copy_data(self, default=None):
...
+            else:
+                this_weekday_number = this_datetime.weekday()
+                if self.delivery_time_preference == "workdays":
+                    if this_weekday_number < 5:
```
Comment (jbaudoux) [inline stock_partner_delivery_window/models/res_partner.py:286]: else missing. You likely need to combine with above test

File: stock_partner_delivery_window/models/res_partner.py [hunk 6]
```diff
@@ -180,3 +183,114 @@ def copy_data(self, default=None):
...
+                        yield this_datetime
+                else:
+                    this_weekday_windows = self.env[
+                        "partner.delivery.time.window"
+                    ].search(
```
Comment (jbaudoux) [inline stock_partner_delivery_window/models/res_partner.py:291]: you should not call search on each iteration, use filtered

File: stock_partner_delivery_window/models/res_partner.py [hunk 7]
```diff
@@ -180,3 +183,114 @@ def copy_data(self, default=None):
...
+                        [
+                            ("time_window_weekday_ids.name", "=", this_weekday_number),
+                            ("partner_id", "=", self.id),
+                        ],
+                        order="time_window_start ASC",
+                    )
+                    for win in this_weekday_windows:
```
Comment (jbaudoux) [inline stock_partner_delivery_window/models/res_partner.py:298]: You should not loop on the time slots. If there is a slot, the day is valid. Next iteration should be on next day.

For now, we can keep time management as a nice to have. Ideally you should check that the time in this_datetime is before the end time of the slot and return the combine of the max(window_stat_time, this_datetime.time). In this case, you also need to reset time to 0 on next iteration (this_datetime += timedelta(days=1))

File: stock_partner_delivery_window/models/res_partner.py [hunk 8]
```diff
@@ -180,3 +183,114 @@ def copy_data(self, default=None):
...
+            ).sorted("time_window_start")
+            for win in this_weekday_windows:
+                this_weekday_start_datetime = datetime.combine(
+                    this_datetime, win.get_time_window_start_time()
+                )
+                res.append(this_weekday_start_datetime)
+        return res
+
+    # TODO: Refactor with function above
+    def get_next_delivery_availability(self, from_datetime, to_datetime=None):
```
Comment (jbaudoux) [inline stock_partner_delivery_window/models/res_partner.py:262]: I would have dropped the `to_datetime`. You can always manage this outside the generator with `myiter.close()`

File: stock_partner_delivery_window/models/res_partner.py [hunk 9]
```diff
@@ -180,3 +183,114 @@ def copy_data(self, default=None):
         ]
         result["delivery_time_window_ids"] = [(0, 0, val) for val in values]
         return [result]
+
+    def next_delivery_window_start_datetime(self, from_date=None, timedelta_days=None):
```
Comment (jbaudoux) [inline stock_partner_delivery_window/models/res_partner.py:189]: Is this still useful or replaced by `get_next_delivery_availability` ?

**Review by jbaudoux — CHANGES_REQUESTED** [review]
I dropped a few comments

**Comment by cyrilmanuel** [comment]
@grindtildeath  little ping, could you made the modification ?

File: stock_partner_delivery_window/models/res_partner.py (hunk 9 above)
Comment (grindtildeath) [reply stock_partner_delivery_window/models/res_partner.py:189]: IMO we can drop this one

File: stock_partner_delivery_window/models/res_partner.py (hunk 4 above)
Comment (grindtildeath) [reply stock_partner_delivery_window/models/res_partner.py]: This is implemented in last commit for delivery windows preference

**Comment by jbaudoux** [comment]
Being replaced by https://github.com/OCA/wms/pull/1026

**Pull request closed without merging**